pkg/controller/constraint/constraint_controller.go

```diff
@@ -274,7 +339,83 @@ func (r *ReconcileConstraint) Reconcile(ctx context.Context, request reconcile.R
+			// generate vapbinding resources
+			if generateVapBinding && IsVapAPIEnabled() {
+				currentVapBinding := &admissionregistrationv1beta1.ValidatingAdmissionPolicyBinding{}
+				vapBindingName := fmt.Sprintf("gatekeeper-%s", instance.GetName())
+				log.Info("check if vapbinding exists", "vapBindingName", vapBindingName)
+				if err := r.reader.Get(ctx, types.NamespacedName{Name: vapBindingName}, currentVapBinding); err != nil {
+					if !apierrors.IsNotFound(err) && !strings.Contains(err.Error(), "failed to get API group resources") {
+						return reconcile.Result{}, err
+					}
+					currentVapBinding = nil
+				}
+				newVapBinding := &admissionregistrationv1beta1.ValidatingAdmissionPolicyBinding{}
+				transformedVapBinding, err := transform.ConstraintToBinding(instance)
+				if err != nil {
+					status.Status.Errors = append(status.Status.Errors, constraintstatusv1beta1.Error{Message: err.Error()})
+					if err2 := r.writer.Update(ctx, status); err2 != nil {
+						log.Error(err2, "could not report transform vapbinding error status")
+					}
+					return reconcile.Result{}, err
+				}
+				if currentVapBinding == nil {
+					newVapBinding = transformedVapBinding.DeepCopy()
+				} else {
+					newVapBinding = currentVapBinding.DeepCopy()
+					newVapBinding.Spec = transformedVapBinding.Spec
+				}
+
+				if err := controllerutil.SetControllerReference(instance, newVapBinding, r.scheme); err != nil {
+					return reconcile.Result{}, err
+				}
+
+				if currentVapBinding == nil {
+					log.Info("creating vapbinding")
+					if err := r.writer.Create(ctx, newVapBinding); err != nil {
+						status.Status.Errors = append(status.Status.Errors, constraintstatusv1beta1.Error{Message: err.Error()})
+						if err2 := r.writer.Update(ctx, status); err2 != nil {
+							log.Error(err2, "could not report creating vapbinding error status")
+						}
+						return reconcile.Result{}, err
+					}
+				}
+				if !reflect.DeepEqual(currentVapBinding, newVapBinding) {
+					log.Info("updating vapbinding")
+					if err := r.writer.Update(ctx, newVapBinding); err != nil {
+						status.Status.Errors = append(status.Status.Errors, constraintstatusv1beta1.Error{Message: err.Error()})
+						if err2 := r.writer.Update(ctx, status); err2 != nil {
+							log.Error(err2, "could not report update vapbinding error status")
+						}
+						return reconcile.Result{}, err
+					}
+				}
+			}
+			// do not generate vapbinding resources
+			// remove if exists
+			if !generateVapBinding && IsVapAPIEnabled() {
+				currentVapBinding := &admissionregistrationv1beta1.ValidatingAdmissionPolicyBinding{}
+				vapBindingName := fmt.Sprintf("gatekeeper-%s", instance.GetName())
+				log.Info("check if vapbinding exists", "vapBindingName", vapBindingName)
+				if err := r.reader.Get(ctx, types.NamespacedName{Name: vapBindingName}, currentVapBinding); err != nil {
+					if !apierrors.IsNotFound(err) && !strings.Contains(err.Error(), "failed to get API group resources") {
```

Comment:
I think you can avoid a string check by running:

`errors.Is(err, ErrResourceDiscoveryFailed)`

trace:

https://github.com/kubernetes-sigs/controller-runtime/blob/56159419231e985c091ef3e7a8a3dee40ddf1d73/pkg/client/apiutil/errors.go#L30-L33C6

https://github.com/kubernetes-sigs/controller-runtime/blob/56159419231e985c091ef3e7a8a3dee40ddf1d73/pkg/client/apiutil/restmapper.go#L303-L306

https://github.com/kubernetes-sigs/controller-runtime/blob/56159419231e985c091ef3e7a8a3dee40ddf1d73/pkg/client/apiutil/restmapper.go#L193-L195

Reply:
Seems https://github.com/kubernetes-sigs/controller-runtime/pull/2472 was added recently and cherry picked to controller-runtime v0.16 but our fork is still on v0.15.

Reply:
de-fork to the rescue :)